examples/lpt_nbody_demo.py

```diff
@@ -0,0 +1,315 @@
+
+  # Apply power spectrum to Fourier modes
+  delta_k *= (power_spectrum(kk) * (nc / box_size)**3)**0.5
+
+  # Compute inverse Fourier transform to recover the initial conditions in real space
+  delta = jaxdecomp.fft.pifft3d(delta_k).real
+
+  # Compute gravitational forces associated with this field
+  grav_kernel = gravity_kernel([kx, ky, kz])
+  forces_k = [g * delta_k for g in grav_kernel]
+
+  # Retrieve the forces in real space by inverse Fourier transforming
+  forces = jnp.stack([jaxdecomp.fft.pifft3d(f).real for f in forces_k], axis=-1)
+
+  return delta, forces
+
+
+def cic_paint(displacement, sharding, halo_size):
+  """ Paints particles on a mesh using Cloud-In-Cell interpolation.
+
+    Parameters
+    ----------
+    displacement : jnp.ndarray
+        Displacement field of particles.
+    sharding : Any
+        Sharding strategy for the array.
+    halo_size : int
+        Halo size for painting.
+
+    Returns
+    -------
+    jnp.ndarray
+        Density field.
+  """
+  local_mesh_shape = _global_to_local_size(displacement.shape, sharding)
+
+  @partial(
+      shard_map,
+      mesh=sharding.mesh,
+      in_specs=(P('z', 'y'),),
+      out_specs=P('z', 'y'))
+  def cic_op(disp):
+    """ CiC operation on each local slice of the mesh."""
+    # Create a mesh to paint the particles on for the local slice
+    mesh = jnp.zeros(disp.shape[:-1], dtype='float32')
+
+    # Padding the output array along the two first dimensions
+    mesh = jnp.pad(mesh,
+                   [[halo_size, halo_size], [halo_size, halo_size], [0, 0]])
+
+    a, b, c = jnp.meshgrid(
+        jnp.arange(local_mesh_shape[0]), jnp.arange(local_mesh_shape[1]),
+        jnp.arange(local_mesh_shape[2]))
+
+    # adding an offset of size halo size
+    pmid = jnp.stack([b + halo_size, a + halo_size, c], axis=-1)
+    return scatter(pmid.reshape([-1, 3]), disp.reshape([-1, 3]), mesh)
+
+  # Performs painting on padded mesh
+  field = cic_op(displacement)
+
+  # Run halo exchange to get the correct values at the boundaries
+  field = jaxdecomp.halo_exchange(
+      field,
+      halo_extents=(halo_size // 2, halo_size // 2, 0),
+      halo_periods=(True, True, True))
+
+  @partial(
+      shard_map,
+      mesh=sharding.mesh,
+      in_specs=(P('z', 'y'),),
+      out_specs=P('z', 'y'))
+  def unpad(x):
+    """ Unpad the output array. """
+    x = x.at[halo_size:halo_size + halo_size // 2].add(x[:halo_size // 2])
+    x = x.at[-(halo_size + halo_size // 2):-halo_size].add(x[-halo_size // 2:])
+    x = x.at[:, halo_size:halo_size + halo_size // 2].add(x[:, :halo_size // 2])
+    x = x.at[:,
+             -(halo_size + halo_size // 2):-halo_size].add(x[:,
+                                                             -halo_size // 2:])
+    return x[halo_size:-halo_size, halo_size:-halo_size, :]
+
+  # Unpad the output array
+  field = unpad(field)
+  return field
+
+
+def simulation_fn(key, nc, box_size, sharding, halo_size, a=1.0):
```

Comment:
This function is not jitted and cannot be jitted.
In general NamedSharding should not be used as an argument (static or not) for a jitted function.
The fact that you used it makes this function unjittable